crates/nix_rs/src/store.rs

```diff
@@ -147,38 +147,41 @@ impl NixStoreCmd {
         &self,
         out_paths: HashSet<StorePath>,
     ) -> Result<HashSet<StorePath>, NixStoreCmdError> {
-        let mut all_drvs = HashSet::new();
-        for out in out_paths.iter() {
-            let drv = self.nix_store_query_deriver(out.clone()).await?;
-            all_drvs.insert(drv);
-        }
-        let mut all_outs = HashSet::new();
-        for drv in all_drvs {
-            let deps = self.nix_store_query_requisites_with_outputs(drv).await?;
-            all_outs.extend(deps.into_iter());
-        }
+        let all_drvs = self
+            .nix_store_query_deriver(&out_paths.iter().cloned().collect::<Vec<_>>())
+            .await?;
+        let all_outs = self
+            .nix_store_query_requisites_with_outputs(&all_drvs.iter().cloned().collect::<Vec<_>>())
+            .await?;
         Ok(all_outs)
     }
 
-    /// Return the derivation used to build the given build output.
+    /// Return the derivations used to build the given build output.
     pub async fn nix_store_query_deriver(
         &self,
-        out_path: StorePath,
-    ) -> Result<PathBuf, NixStoreCmdError> {
+        out_paths: &[StorePath],
+    ) -> Result<HashSet<PathBuf>, NixStoreCmdError> {
```

Comment:
This should return `Vec` (so as to maintain original order).

Also, I see no reason for it to require a HashSet over Vec. The convenience of a single caller should not universally dictate API of some core function.

Reply:
Even I thought about this but then chose HashSet thinking about consistency as main function `fetch_all_deps` also returns a HashSet and the other reason was HashSet would contain unique while Vec won't.

Reply:
But thinking about this `The convenience of a single caller should not universally dictate API of some core function.` now I am thinking I can collect the final result in a HashSet and return.